src/core/markovProcessGraph.test.js

```diff
@@ -0,0 +1,428 @@
+    new Graph()
+      .addNode(participantNode)
+      .addNode(c0)
+      .addNode(c1)
+      .addEdge(e0)
+      .addEdge(e1)
+      .addEdge(e2);
+  const nodeWeights = () => new Map().set(c0.address, 1).set(c1.address, 2);
+  const edgeWeights = () =>
+    new Map()
+      .set(e0.address, {forwards: 1, backwards: 0})
+      .set(e1.address, {forwards: 1, backwards: 1});
+  const weights = () => ({
+    nodeWeights: nodeWeights(),
+    edgeWeights: edgeWeights(),
+  });
+  const weightedGraph = () => ({weights: weights(), graph: graph()});
+  const args = () => ({
+    weightedGraph: weightedGraph(),
+    parameters,
+    intervals,
+    participants: [participant],
+  });
+  const markovProcessGraph = () => MarkovProcessGraph.new(args());
+
+  function checkMarkovEdge(mpg: MarkovProcessGraph, me: MarkovEdge) {
+    const addr = markovEdgeAddressFromMarkovEdge(me);
+    const actual = mpg.edge(addr);
+    if (actual == null) {
+      throw new Error(
+        `no markov edge matching ${MarkovEdgeAddress.toString(addr)}`
+      );
+    }
+    expect({...me, transitionProbability: expect.anything()}).toEqual(actual);
+    expect(me.transitionProbability).toBeCloseTo(actual.transitionProbability);
+  }
+
+  describe("basic validation", () => {
+    it("errors for negative parameters", () => {
+      const badParameters = [
+        {alpha: -0.1, beta: 0.1, gammaForward: 0.1, gammaBackward: 0.1},
+        {alpha: 0.1, beta: -0.1, gammaForward: 0.1, gammaBackward: 0.1},
+        {alpha: 0.1, beta: 0.1, gammaForward: -0.1, gammaBackward: 0.1},
+        {alpha: 0.1, beta: 0.1, gammaForward: 0.1, gammaBackward: -0.1},
+      ];
+      for (const b of badParameters) {
+        const badArgs = {...args(), parameters: b};
+        const thunk = () => MarkovProcessGraph.new(badArgs);
+        expect(thunk).toThrowError("Invalid transition probability");
+      }
+    });
+    it("errors if parameters sum greater than 1", () => {
+      const bad = {
+        alpha: 0.25,
+        gammaForward: 0.25,
+        gammaBackward: 0.3,
+        beta: 0.25,
+      };
+      const thunk = () => MarkovProcessGraph.new({...args(), parameters: bad});
+      expect(thunk).toThrowError("Overlarge transition probability");
+    });
+  });
+
+  describe("organic nodes / edges", () => {
+    it("each contribution is present with the correct weight", () => {
+      const mpg = markovProcessGraph();
+      const n0 = NullUtil.get(mpg.node(c0.address));
+      expect(n0.mint).toEqual(nodeWeights().get(n0.address));
+      const n1 = NullUtil.get(mpg.node(c1.address));
+      expect(n1.mint).toEqual(nodeWeights().get(n1.address));
+    });
+    it("the participant is not present", () => {
+      const mpg = markovProcessGraph();
+      expect(mpg.node(participant.address)).toEqual(null);
+    });
+    it("the edge between contributions corresponds to two MarkovEdges", () => {
+      const mpg = markovProcessGraph();
+      const aF = markovEdgeAddress(e2.address, "F");
```

Comment:
e0 is unusual since it get's rewritten. i added a new edge, e3, with zero weight in both directions, to add an explicit test case for the filtering of 0-weight edges.

Reply:
Lovely; thanks!